broker/src/test/java/io/camunda/zeebe/broker/transport/commandapi/CommandApiRequestHandlerTest.java

```diff
@@ -0,0 +1,140 @@
+  final CommandApiRequestHandler handler = new CommandApiRequestHandler();
+
+  @Before
+  public void setup() {
+    scheduler.submitActor(handler);
+    handler.addPartition(0, mock(LogStreamRecordWriter.class), new NoopRequestLimiter<>());
+    scheduler.workUntilDone();
+  }
+
+  @Test
+  public void shouldRejectCommandIfNotLeader() {
+    // given
+    final var request = new ExecuteCommandRequest();
+    handler.removePartition(0);
+    scheduler.workUntilDone();
+
+    // when
+    final var response = handleRequest(request);
+
+    // then
+    assertThat(response).matches(Either::isLeft);
+    assertThat(response.getLeft().getErrorCode()).isEqualTo(ErrorCode.PARTITION_LEADER_MISMATCH);
+  }
+
+  @Test
+  public void shouldRejectCommandWithoutEvent() {
+    // given
+    final var request = new ExecuteCommandRequest();
+
+    // when
+    final var response = handleRequest(request);
+
+    // then
+    assertThat(response).matches(Either::isLeft);
+    assertThat(response.getLeft().getErrorCode()).isEqualTo(ErrorCode.UNSUPPORTED_MESSAGE);
+  }
+
+  @Test
+  public void shouldRejectCommandWithUnknownEvent() {
+    // given
+    final var request = new ExecuteCommandRequest();
+    request.setValueType(ValueType.ERROR);
+
+    // when
+    final var response = handleRequest(request);
+
+    // then
+    assertThat(response).matches(Either::isLeft);
+    assertThat(response.getLeft().getErrorCode()).isEqualTo(ErrorCode.UNSUPPORTED_MESSAGE);
+  }
+
+  @Test
+  public void shouldRejectCommandIfResourcesExhausted() {
+    // given
+    final RequestLimiter<Intent> limiter = mock(RequestLimiter.class);
+    when(limiter.tryAcquire(anyInt(), anyLong(), any())).thenReturn(false);
+    handler.addPartition(0, mock(LogStreamRecordWriter.class), limiter);
+    scheduler.workUntilDone();
+
+    final var request =
+        new BrokerPublishMessageRequest("test", "1").setMessageId("1").setTimeToLive(0);
+    request.serializeValue();
+
+    // when
+    final var response = handleRequest(request);
+
+    // then
+    assertThat(response).matches(Either::isLeft);
+    assertThat(response.getLeft().getErrorCode()).isEqualTo(ErrorCode.RESOURCE_EXHAUSTED);
+  }
+
+  @Test
+  public void shouldWriteToLog() {
+    // given
+    final var logWriter = mock(LogStreamRecordWriter.class);
+    when(logWriter.metadataWriter(any())).thenReturn(logWriter);
+    when(logWriter.valueWriter(any())).thenReturn(logWriter);
+    handler.addPartition(0, logWriter, new NoopRequestLimiter<>());
+    scheduler.workUntilDone();
+
+    final var request =
+        new BrokerPublishMessageRequest("test", "1").setMessageId("1").setTimeToLive(0);
+    request.serializeValue();
+
+    // when
+    final var response = handleRequest(request);
+
+    // then
+    assertThat(response).matches(Either::isRight);
+    verify(logWriter).tryWrite();
+    verify(logWriter).reset();
+  }
+
+  private Either<ErrorResponseWriter, CommandApiResponseWriter> handleRequest(
```

Comment:
This is not executed by the actor. Also requires exposing an internal method as public. Can we test it as a black-box similar to QueryApiTest? If it is too much effort to do it, you can ignore this. But I would assume this is easy to do it as you can copy the setup from QueryApiTest. We are only interested in ErrorResponse, so you only have to decode the error response.

Reply:
I've copied over some code from the query api to directly test `onRequest` instead of the internal `handleExecuteCommandRequest`